### DISCUSSION TITLE
RetryTopicConfiguration doesn't work if KafkaListener is used on class level

### DISCUSSION THREAD

**Discussion by Casper-dss** [spring-projects/spring-kafka#3010] [Q&A] [answered]
Hi

I am creating non-blocking retry through RetryTopicConfiguration. It works fine if I have a KafkaListener on method level. But if I move it to class level and use a KafkaHandler on method level, it will use the DefaultErrorHandler backoff. Can't I use RetryTopicConfiguration together with class level KafkaListener?

**Comment by danipenaperez** [comment]
Hi @Casper-dss I was in the same trouble (spring-kafka 2.9.X) . 
After diving (and suffering)  I found this restricticion in source code:
`https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java#L282`
Let's see if someone could clarify this behavior

**Comment by sobychacko** [comment]
The reference docs also make it clear that `KafkaListener` is not supported at the class-level at the moment when using the nonblocking retries. https://docs.spring.io/spring-kafka/reference/retrytopic/how-the-pattern-works.html.

**Reply by Casper-dss** [reply]
Sorry I didn't see it. Is a workaround available?

**Reply by sobychacko** [reply, marked as answer]
Currently, you have to use `KafkaListener` at the method level. The complexity of the changes involved is why this may not be supported. Feel free to create an issue in the backlog, and we will get to it, or someone from the community can address it and contribute a fix.
